// @ctrl/tinycolor@3.6.1/dist/module/format-input/index.js
var F = Object.defineProperty;
var x = (a, t) => F(a, "name", { value: t, configurable: !0 });
import { rgbToRgb as R, hsvToRgb as E, hslToRgb as T, parseIntFromHex as n, convertHexToDecimal as S } from "../conversion/index.js";
import { names as A } from "../css-color-names/index.js";
import { convertToPercentage as m, boundAlpha as M } from "../util/index.js";
function y(a) {
  var t = { r: 0, g: 0, b: 0 }, e = 1, h = null, u = null, p = null, g = !1, i = !1;
  return typeof a == "string" && (a = C(a)), typeof a == "object" && (f(a.r) && f(a.g) && f(a.b) ? (t = R(a.r, a.g, a.b), g = !0, i = String(a.r).substr(-1) === "%" ? "prgb" : "rgb") : f(a.h) && f(a.s) && f(a.v) ? (h = m(a.s), u = m(a.v), t = E(a.h, h, u), g = !0, i = "hsv") : f(a.h) && f(a.s) && f(a.l) && (h = m(a.s), p = m(a.l), t = T(a.h, h, p), g = !0, i = "hsl"), Object.prototype.hasOwnProperty.call(a, "a") && (e = a.a)), e = M(e), {
    ok: g,
    format: a.format || i,
    r: Math.min(255, Math.max(t.r, 0)),
    g: Math.min(255, Math.max(t.g, 0)),
    b: Math.min(255, Math.max(t.b, 0)),
    a: e
  };
}
x(y, "inputToRGB");
var I = "[-\\+]?\\d+%?", w = "[-\\+]?\\d*\\.\\d+%?", s = "(?:".concat(w, ")|(?:").concat(I, ")"), b = "[\\s|\\(]+(".concat(s, ")[,|\\s]+(").concat(s, ")[,|\\s]+(").concat(s, ")\\s*\\)?"), v = "[\\s|\\(]+(".concat(s, ")[,|\\s]+(").concat(s, ")[,|\\s]+(").concat(s, ")[,|\\s]+(").concat(s, ")\\s*\\)?"), r = {
  CSS_UNIT: new RegExp(s),
  rgb: new RegExp("rgb" + b),
  rgba: new RegExp("rgba" + v),
  hsl: new RegExp("hsl" + b),
  hsla: new RegExp("hsla" + v),
  hsv: new RegExp("hsv" + b),
  hsva: new RegExp("hsva" + v),
  hex3: /^#?([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex6: /^#?([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/,
  hex4: /^#?([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex8: /^#?([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/
};
function C(a) {
  if (a = a.trim().toLowerCase(), a.length === 0)
    return !1;
  var t = !1;
  if (A[a])
    a = A[a], t = !0;
  else if (a === "transparent")
    return { r: 0, g: 0, b: 0, a: 0, format: "name" };
  var e = r.rgb.exec(a);
  return e ? { r: e[1], g: e[2], b: e[3] } : (e = r.rgba.exec(a), e ? { r: e[1], g: e[2], b: e[3], a: e[4] } : (e = r.hsl.exec(a), e ? { h: e[1], s: e[2], l: e[3] } : (e = r.hsla.exec(a), e ? { h: e[1], s: e[2], l: e[3], a: e[4] } : (e = r.hsv.exec(a), e ? { h: e[1], s: e[2], v: e[3] } : (e = r.hsva.exec(a), e ? { h: e[1], s: e[2], v: e[3], a: e[4] } : (e = r.hex8.exec(a), e ? {
    r: n(e[1]),
    g: n(e[2]),
    b: n(e[3]),
    a: S(e[4]),
    format: t ? "name" : "hex8"
  } : (e = r.hex6.exec(a), e ? {
    r: n(e[1]),
    g: n(e[2]),
    b: n(e[3]),
    format: t ? "name" : "hex"
  } : (e = r.hex4.exec(a), e ? {
    r: n(e[1] + e[1]),
    g: n(e[2] + e[2]),
    b: n(e[3] + e[3]),
    a: S(e[4] + e[4]),
    format: t ? "name" : "hex8"
  } : (e = r.hex3.exec(a), e ? {
    r: n(e[1] + e[1]),
    g: n(e[2] + e[2]),
    b: n(e[3] + e[3]),
    format: t ? "name" : "hex"
  } : !1)))))))));
}
x(C, "stringInputToObject");
function f(a) {
  return !!r.CSS_UNIT.exec(String(a));
}
x(f, "isValidCSSUnit");
export {
  y as inputToRGB,
  f as isValidCSSUnit,
  C as stringInputToObject
};
