// dayjs@1.11.13/dayjs.min/index.js
var R = Object.defineProperty;
var d = (A, J) => R(A, "name", { value: J, configurable: !0 });
import { commonjsGlobal as X, getDefaultExportFromCjs as tt } from "../../../../../../_virtual/_commonjsHelpers/index.js";
import { __module as B } from "../../../../../../_virtual/dayjs.min/index.js";
(function(A, J) {
  (function(I, x) {
    A.exports = x();
  })(X, function() {
    var I = 1e3, x = 6e4, Z = 36e5, F = "millisecond", w = "second", O = "minute", _ = "hour", y = "day", j = "week", M = "month", z = "quarter", v = "year", b = "date", E = "Invalid Date", G = /^(\d{4})[-/]?(\d{1,2})?[-/]?(\d{0,2})[Tt\s]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?[.:]?(\d+)?$/, P = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g, Q = { name: "en", weekdays: "Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday".split("_"), months: "January_February_March_April_May_June_July_August_September_October_November_December".split("_"), ordinal: /* @__PURE__ */ d(function(s) {
      var n = ["th", "st", "nd", "rd"], t = s % 100;
      return "[" + s + (n[(t - 20) % 10] || n[t] || n[0]) + "]";
    }, "ordinal") }, N = /* @__PURE__ */ d(function(s, n, t) {
      var r = String(s);
      return !r || r.length >= n ? s : "" + Array(n + 1 - r.length).join(t) + s;
    }, "m"), K = { s: N, z: /* @__PURE__ */ d(function(s) {
      var n = -s.utcOffset(), t = Math.abs(n), r = Math.floor(t / 60), e = t % 60;
      return (n <= 0 ? "+" : "-") + N(r, 2, "0") + ":" + N(e, 2, "0");
    }, "z"), m: /* @__PURE__ */ d(function s(n, t) {
      if (n.date() < t.date()) return -s(t, n);
      var r = 12 * (t.year() - n.year()) + (t.month() - n.month()), e = n.clone().add(r, M), i = t - e < 0, u = n.clone().add(r + (i ? -1 : 1), M);
      return +(-(r + (t - e) / (i ? e - u : u - e)) || 0);
    }, "t"), a: /* @__PURE__ */ d(function(s) {
      return s < 0 ? Math.ceil(s) || 0 : Math.floor(s);
    }, "a"), p: /* @__PURE__ */ d(function(s) {
      return { M, y: v, w: j, d: y, D: b, h: _, m: O, s: w, ms: F, Q: z }[s] || String(s || "").toLowerCase().replace(/s$/, "");
    }, "p"), u: /* @__PURE__ */ d(function(s) {
      return s === void 0;
    }, "u") }, Y = "en", g = {};
    g[Y] = Q;
    var V = "$isDayjsObject", U = /* @__PURE__ */ d(function(s) {
      return s instanceof L || !(!s || !s[V]);
    }, "S"), C = /* @__PURE__ */ d(function s(n, t, r) {
      var e;
      if (!n) return Y;
      if (typeof n == "string") {
        var i = n.toLowerCase();
        g[i] && (e = i), t && (g[i] = t, e = i);
        var u = n.split("-");
        if (!e && u.length > 1) return s(u[0]);
      } else {
        var o = n.name;
        g[o] = n, e = o;
      }
      return !r && e && (Y = e), e || !r && Y;
    }, "t"), f = /* @__PURE__ */ d(function(s, n) {
      if (U(s)) return s.clone();
      var t = typeof n == "object" ? n : {};
      return t.date = s, t.args = arguments, new L(t);
    }, "O"), a = K;
    a.l = C, a.i = U, a.w = function(s, n) {
      return f(s, { locale: n.$L, utc: n.$u, x: n.$x, $offset: n.$offset });
    };
    var L = function() {
      function s(t) {
        this.$L = C(t.locale, null, !0), this.parse(t), this.$x = this.$x || t.x || {}, this[V] = !0;
      }
      d(s, "M");
      var n = s.prototype;
      return n.parse = function(t) {
        this.$d = function(r) {
          var e = r.date, i = r.utc;
          if (e === null) return /* @__PURE__ */ new Date(NaN);
          if (a.u(e)) return /* @__PURE__ */ new Date();
          if (e instanceof Date) return new Date(e);
          if (typeof e == "string" && !/Z$/i.test(e)) {
            var u = e.match(G);
            if (u) {
              var o = u[2] - 1 || 0, c = (u[7] || "0").substring(0, 3);
              return i ? new Date(Date.UTC(u[1], o, u[3] || 1, u[4] || 0, u[5] || 0, u[6] || 0, c)) : new Date(u[1], o, u[3] || 1, u[4] || 0, u[5] || 0, u[6] || 0, c);
            }
          }
          return new Date(e);
        }(t), this.init();
      }, n.init = function() {
        var t = this.$d;
        this.$y = t.getFullYear(), this.$M = t.getMonth(), this.$D = t.getDate(), this.$W = t.getDay(), this.$H = t.getHours(), this.$m = t.getMinutes(), this.$s = t.getSeconds(), this.$ms = t.getMilliseconds();
      }, n.$utils = function() {
        return a;
      }, n.isValid = function() {
        return this.$d.toString() !== E;
      }, n.isSame = function(t, r) {
        var e = f(t);
        return this.startOf(r) <= e && e <= this.endOf(r);
      }, n.isAfter = function(t, r) {
        return f(t) < this.startOf(r);
      }, n.isBefore = function(t, r) {
        return this.endOf(r) < f(t);
      }, n.$g = function(t, r, e) {
        return a.u(t) ? this[r] : this.set(e, t);
      }, n.unix = function() {
        return Math.floor(this.valueOf() / 1e3);
      }, n.valueOf = function() {
        return this.$d.getTime();
      }, n.startOf = function(t, r) {
        var e = this, i = !!a.u(r) || r, u = a.p(t), o = /* @__PURE__ */ d(function(S, l) {
          var D = a.w(e.$u ? Date.UTC(e.$y, l, S) : new Date(e.$y, l, S), e);
          return i ? D : D.endOf(y);
        }, "l"), c = /* @__PURE__ */ d(function(S, l) {
          return a.w(e.toDate()[S].apply(e.toDate("s"), (i ? [0, 0, 0, 0] : [23, 59, 59, 999]).slice(l)), e);
        }, "$"), h = this.$W, $ = this.$M, m = this.$D, k = "set" + (this.$u ? "UTC" : "");
        switch (u) {
          case v:
            return i ? o(1, 0) : o(31, 11);
          case M:
            return i ? o(1, $) : o(0, $ + 1);
          case j:
            var p = this.$locale().weekStart || 0, H = (h < p ? h + 7 : h) - p;
            return o(i ? m - H : m + (6 - H), $);
          case y:
          case b:
            return c(k + "Hours", 0);
          case _:
            return c(k + "Minutes", 1);
          case O:
            return c(k + "Seconds", 2);
          case w:
            return c(k + "Milliseconds", 3);
          default:
            return this.clone();
        }
      }, n.endOf = function(t) {
        return this.startOf(t, !1);
      }, n.$set = function(t, r) {
        var e, i = a.p(t), u = "set" + (this.$u ? "UTC" : ""), o = (e = {}, e[y] = u + "Date", e[b] = u + "Date", e[M] = u + "Month", e[v] = u + "FullYear", e[_] = u + "Hours", e[O] = u + "Minutes", e[w] = u + "Seconds", e[F] = u + "Milliseconds", e)[i], c = i === y ? this.$D + (r - this.$W) : r;
        if (i === M || i === v) {
          var h = this.clone().set(b, 1);
          h.$d[o](c), h.init(), this.$d = h.set(b, Math.min(this.$D, h.daysInMonth())).$d;
        } else o && this.$d[o](c);
        return this.init(), this;
      }, n.set = function(t, r) {
        return this.clone().$set(t, r);
      }, n.get = function(t) {
        return this[a.p(t)]();
      }, n.add = function(t, r) {
        var e, i = this;
        t = Number(t);
        var u = a.p(r), o = /* @__PURE__ */ d(function($) {
          var m = f(i);
          return a.w(m.date(m.date() + Math.round($ * t)), i);
        }, "y");
        if (u === M) return this.set(M, this.$M + t);
        if (u === v) return this.set(v, this.$y + t);
        if (u === y) return o(1);
        if (u === j) return o(7);
        var c = (e = {}, e[O] = x, e[_] = Z, e[w] = I, e)[u] || 1, h = this.$d.getTime() + t * c;
        return a.w(h, this);
      }, n.subtract = function(t, r) {
        return this.add(-1 * t, r);
      }, n.format = function(t) {
        var r = this, e = this.$locale();
        if (!this.isValid()) return e.invalidDate || E;
        var i = t || "YYYY-MM-DDTHH:mm:ssZ", u = a.z(this), o = this.$H, c = this.$m, h = this.$M, $ = e.weekdays, m = e.months, k = e.meridiem, p = /* @__PURE__ */ d(function(l, D, T, W) {
          return l && (l[D] || l(r, i)) || T[D].slice(0, W);
        }, "h"), H = /* @__PURE__ */ d(function(l) {
          return a.s(o % 12 || 12, l, "0");
        }, "d"), S = k || function(l, D, T) {
          var W = l < 12 ? "AM" : "PM";
          return T ? W.toLowerCase() : W;
        };
        return i.replace(P, function(l, D) {
          return D || function(T) {
            switch (T) {
              case "YY":
                return String(r.$y).slice(-2);
              case "YYYY":
                return a.s(r.$y, 4, "0");
              case "M":
                return h + 1;
              case "MM":
                return a.s(h + 1, 2, "0");
              case "MMM":
                return p(e.monthsShort, h, m, 3);
              case "MMMM":
                return p(m, h);
              case "D":
                return r.$D;
              case "DD":
                return a.s(r.$D, 2, "0");
              case "d":
                return String(r.$W);
              case "dd":
                return p(e.weekdaysMin, r.$W, $, 2);
              case "ddd":
                return p(e.weekdaysShort, r.$W, $, 3);
              case "dddd":
                return $[r.$W];
              case "H":
                return String(o);
              case "HH":
                return a.s(o, 2, "0");
              case "h":
                return H(1);
              case "hh":
                return H(2);
              case "a":
                return S(o, c, !0);
              case "A":
                return S(o, c, !1);
              case "m":
                return String(c);
              case "mm":
                return a.s(c, 2, "0");
              case "s":
                return String(r.$s);
              case "ss":
                return a.s(r.$s, 2, "0");
              case "SSS":
                return a.s(r.$ms, 3, "0");
              case "Z":
                return u;
            }
            return null;
          }(l) || u.replace(":", "");
        });
      }, n.utcOffset = function() {
        return 15 * -Math.round(this.$d.getTimezoneOffset() / 15);
      }, n.diff = function(t, r, e) {
        var i, u = this, o = a.p(r), c = f(t), h = (c.utcOffset() - this.utcOffset()) * x, $ = this - c, m = /* @__PURE__ */ d(function() {
          return a.m(u, c);
        }, "D");
        switch (o) {
          case v:
            i = m() / 12;
            break;
          case M:
            i = m();
            break;
          case z:
            i = m() / 3;
            break;
          case j:
            i = ($ - h) / 6048e5;
            break;
          case y:
            i = ($ - h) / 864e5;
            break;
          case _:
            i = $ / Z;
            break;
          case O:
            i = $ / x;
            break;
          case w:
            i = $ / I;
            break;
          default:
            i = $;
        }
        return e ? i : a.a(i);
      }, n.daysInMonth = function() {
        return this.endOf(M).$D;
      }, n.$locale = function() {
        return g[this.$L];
      }, n.locale = function(t, r) {
        if (!t) return this.$L;
        var e = this.clone(), i = C(t, r, !0);
        return i && (e.$L = i), e;
      }, n.clone = function() {
        return a.w(this.$d, this);
      }, n.toDate = function() {
        return new Date(this.valueOf());
      }, n.toJSON = function() {
        return this.isValid() ? this.toISOString() : null;
      }, n.toISOString = function() {
        return this.$d.toISOString();
      }, n.toString = function() {
        return this.$d.toUTCString();
      }, s;
    }(), q = L.prototype;
    return f.prototype = q, [["$ms", F], ["$s", w], ["$m", O], ["$H", _], ["$W", y], ["$M", M], ["$y", v], ["$D", b]].forEach(function(s) {
      q[s[1]] = function(n) {
        return this.$g(n, s[0], s[1]);
      };
    }), f.extend = function(s, n) {
      return s.$i || (s(n, L, f), s.$i = !0), f;
    }, f.locale = C, f.isDayjs = U, f.unix = function(s) {
      return f(1e3 * s);
    }, f.en = g[Y], f.Ls = g, f.p = {}, f;
  });
})(B);
var et = B.exports;
const it = /* @__PURE__ */ tt(et);
export {
  it as default
};
