// element-plus@2.8.1/es/hooks/use-locale/index/index.js
var u = Object.defineProperty;
var n = (t, e) => u(t, "name", { value: e, configurable: !0 });
import { inject as m, ref as c, computed as s, unref as a, isRef as i } from "vue";
import f from "../../../locale/lang/en/index.js";
import d from "../../../../../../../lodash-es@4.17.21/node_modules/lodash-es/get/index.js";
const g = /* @__PURE__ */ n((t) => (e, o) => p(e, o, a(t)), "buildTranslator"), p = /* @__PURE__ */ n((t, e, o) => d(o, t, t).replace(/\{(\w+)\}/g, (v, r) => {
  var l;
  return `${(l = e == null ? void 0 : e[r]) != null ? l : `{${r}}`}`;
}), "translate"), x = /* @__PURE__ */ n((t) => {
  const e = s(() => a(t).name), o = i(t) ? t : c(t);
  return {
    lang: e,
    locale: o,
    t: g(t)
  };
}, "buildLocaleContext"), b = Symbol("localeContextKey"), R = /* @__PURE__ */ n((t) => {
  const e = t || m(b, c());
  return x(s(() => e.value || f));
}, "useLocale");
export {
  x as buildLocaleContext,
  g as buildTranslator,
  b as localeContextKey,
  p as translate,
  R as useLocale
};
