// element-plus@2.8.1/es/components/cascader-panel/src/config/index.js
var a = Object.defineProperty;
var e = (r, o) => a(r, "name", { value: o, configurable: !0 });
import { computed as t } from "vue";
import { NOOP as p } from "../../../../../../../../@vue_shared@3.5.13/node_modules/@vue/shared/dist/shared.esm-bundler/index.js";
import { buildProps as i, definePropType as l } from "../../../../utils/vue/props/runtime/index.js";
const u = i({
  modelValue: {
    type: l([Number, String, Array])
  },
  options: {
    type: l(Array),
    default: /* @__PURE__ */ e(() => [], "default")
  },
  props: {
    type: l(Object),
    default: /* @__PURE__ */ e(() => ({}), "default")
  }
}), s = {
  expandTrigger: "click",
  multiple: !1,
  checkStrictly: !1,
  emitPath: !0,
  lazy: !1,
  lazyLoad: p,
  value: "value",
  label: "label",
  children: "children",
  leaf: "leaf",
  disabled: "disabled",
  hoverThreshold: 500
}, n = /* @__PURE__ */ e((r) => t(() => ({
  ...s,
  ...r.props
})), "useCascaderConfig");
export {
  u as CommonProps,
  s as DefaultProps,
  n as useCascaderConfig
};
