// dayjs@1.11.13/plugin/customParseFormat/index.js
var ot = Object.defineProperty;
var L = (F, Q) => ot(F, "name", { value: Q, configurable: !0 });
import { commonjsGlobal as at, getDefaultExportFromCjs as st } from "../../../../../../../_virtual/_commonjsHelpers/index.js";
import { __module as K } from "../../../../../../../_virtual/customParseFormat/index.js";
(function(F, Q) {
  (function(X, H) {
    F.exports = H();
  })(at, function() {
    var X = { LTS: "h:mm:ss A", LT: "h:mm A", L: "MM/DD/YYYY", LL: "MMMM D, YYYY", LLL: "MMMM D, YYYY h:mm A", LLLL: "dddd, MMMM D, YYYY h:mm A" }, H = /(\[[^[]*\])|([-_:/.,()\s]+)|(A|a|Q|YYYY|YY?|ww?|MM?M?M?|Do|DD?|hh?|HH?|mm?|ss?|S{1,3}|z|ZZ?)/g, G = /\d/, x = /\d\d/, s = /\d\d?/, y = /\d*[^-_:/,()\s\d]+/, m = {}, V = /* @__PURE__ */ L(function(t) {
      return (t = +t) + (t > 68 ? 1900 : 2e3);
    }, "a"), o = /* @__PURE__ */ L(function(t) {
      return function(r) {
        this[t] = +r;
      };
    }, "f"), q = [/[+-]\d\d:?(\d\d)?|Z/, function(t) {
      (this.zone || (this.zone = {})).offset = function(r) {
        if (!r || r === "Z") return 0;
        var e = r.match(/([+-]|\d\d)/g), n = 60 * e[1] + (+e[2] || 0);
        return n === 0 ? 0 : e[0] === "+" ? -n : n;
      }(t);
    }], O = /* @__PURE__ */ L(function(t) {
      var r = m[t];
      return r && (r.indexOf ? r : r.s.concat(r.f));
    }, "u"), B = /* @__PURE__ */ L(function(t, r) {
      var e, n = m.meridiem;
      if (n) {
        for (var h = 1; h <= 24; h += 1) if (t.indexOf(n(h, 0, r)) > -1) {
          e = h > 12;
          break;
        }
      } else e = t === (r ? "pm" : "PM");
      return e;
    }, "d"), N = { A: [y, function(t) {
      this.afternoon = B(t, !1);
    }], a: [y, function(t) {
      this.afternoon = B(t, !0);
    }], Q: [G, function(t) {
      this.month = 3 * (t - 1) + 1;
    }], S: [G, function(t) {
      this.milliseconds = 100 * +t;
    }], SS: [x, function(t) {
      this.milliseconds = 10 * +t;
    }], SSS: [/\d{3}/, function(t) {
      this.milliseconds = +t;
    }], s: [s, o("seconds")], ss: [s, o("seconds")], m: [s, o("minutes")], mm: [s, o("minutes")], H: [s, o("hours")], h: [s, o("hours")], HH: [s, o("hours")], hh: [s, o("hours")], D: [s, o("day")], DD: [x, o("day")], Do: [y, function(t) {
      var r = m.ordinal, e = t.match(/\d+/);
      if (this.day = e[0], r) for (var n = 1; n <= 31; n += 1) r(n).replace(/\[|\]/g, "") === t && (this.day = n);
    }], w: [s, o("week")], ww: [x, o("week")], M: [s, o("month")], MM: [x, o("month")], MMM: [y, function(t) {
      var r = O("months"), e = (O("monthsShort") || r.map(function(n) {
        return n.slice(0, 3);
      })).indexOf(t) + 1;
      if (e < 1) throw new Error();
      this.month = e % 12 || e;
    }], MMMM: [y, function(t) {
      var r = O("months").indexOf(t) + 1;
      if (r < 1) throw new Error();
      this.month = r % 12 || r;
    }], Y: [/[+-]?\d+/, o("year")], YY: [x, function(t) {
      this.year = V(t);
    }], YYYY: [/\d{4}/, o("year")], Z: q, ZZ: q };
    function R(t) {
      var r, e;
      r = t, e = m && m.formats;
      for (var n = (t = r.replace(/(\[[^\]]+])|(LTS?|l{1,4}|L{1,4})/g, function(v, l, f) {
        var i = f && f.toUpperCase();
        return l || e[f] || X[f] || e[i].replace(/(\[[^\]]+])|(MMMM|MM|DD|dddd)/g, function(M, Y, D) {
          return Y || D.slice(1);
        });
      })).match(H), h = n.length, c = 0; c < h; c += 1) {
        var $ = n[c], p = N[$], u = p && p[0], d = p && p[1];
        n[c] = d ? { regex: u, parser: d } : $.replace(/^\[|\]$/g, "");
      }
      return function(v) {
        for (var l = {}, f = 0, i = 0; f < h; f += 1) {
          var M = n[f];
          if (typeof M == "string") i += M.length;
          else {
            var Y = M.regex, D = M.parser, S = v.slice(i), g = Y.exec(S)[0];
            D.call(l, g), v = v.replace(g, "");
          }
        }
        return function(w) {
          var A = w.afternoon;
          if (A !== void 0) {
            var a = w.hours;
            A ? a < 12 && (w.hours += 12) : a === 12 && (w.hours = 0), delete w.afternoon;
          }
        }(l), l;
      };
    }
    return L(R, "l"), function(t, r, e) {
      e.p.customParseFormat = !0, t && t.parseTwoDigitYear && (V = t.parseTwoDigitYear);
      var n = r.prototype, h = n.parse;
      n.parse = function(c) {
        var $ = c.date, p = c.utc, u = c.args;
        this.$u = p;
        var d = u[1];
        if (typeof d == "string") {
          var v = u[2] === !0, l = u[3] === !0, f = v || l, i = u[2];
          l && (i = u[2]), m = this.$locale(), !v && i && (m = e.Ls[i]), this.$d = function(S, g, w, A) {
            try {
              if (["x", "X"].indexOf(g) > -1) return new Date((g === "X" ? 1e3 : 1) * S);
              var a = R(g)(S), P = a.year, T = a.month, W = a.day, tt = a.hours, rt = a.minutes, et = a.seconds, nt = a.milliseconds, I = a.zone, J = a.week, z = /* @__PURE__ */ new Date(), C = W || (P || T ? 1 : z.getDate()), E = P || z.getFullYear(), Z = 0;
              P && !T || (Z = T > 0 ? T - 1 : z.getMonth());
              var k, _ = tt || 0, b = rt || 0, U = et || 0, j = nt || 0;
              return I ? new Date(Date.UTC(E, Z, C, _, b, U, j + 60 * I.offset * 1e3)) : w ? new Date(Date.UTC(E, Z, C, _, b, U, j)) : (k = new Date(E, Z, C, _, b, U, j), J && (k = A(k).week(J).toDate()), k);
            } catch {
              return /* @__PURE__ */ new Date("");
            }
          }($, d, p, e), this.init(), i && i !== !0 && (this.$L = this.locale(i).$L), f && $ != this.format(d) && (this.$d = /* @__PURE__ */ new Date("")), m = {};
        } else if (d instanceof Array) for (var M = d.length, Y = 1; Y <= M; Y += 1) {
          u[1] = d[Y - 1];
          var D = e.apply(this, u);
          if (D.isValid()) {
            this.$d = D.$d, this.$L = D.$L, this.init();
            break;
          }
          Y === M && (this.$d = /* @__PURE__ */ new Date(""));
        }
        else h.call(this, c);
      };
    };
  });
})(K);
var it = K.exports;
const dt = /* @__PURE__ */ st(it);
export {
  dt as default
};
